tests/unit/api/objectPut.js

```diff
@@ -269,6 +429,30 @@ describe('objectPut API', () => {
                 });
         });
     });
+
+    it('should not put object with retention configuration if object lock ' +
+        'is not enabled on the bucket', done => {
+        const testPutObjectRequest = new DummyRequest({
+            bucketName,
+            namespace,
+            objectKey: objectName,
+            headers: {
+                'x-amz-object-lock-retain-until-date': mockDate,
+                'x-amz-object-lock-mode': 'GOVERNANCE',
+            },
+            url: `/${bucketName}/${objectName}`,
+            calculatedHash: 'vnR+tLdVF79rPPfF+7YvOg==',
+        }, postBody);
+
+        bucketPut(authInfo, testPutBucketRequest, log, () => {
+            objectPut(authInfo, testPutObjectRequest, undefined, log, err => {
+                assert.deepStrictEqual(err, errors.InvalidRequest
```

Comment:
String comparing the error description is probably a bit much. Altering descriptions like this should not typically break tests.

Reply:
(seems like I missed this one earlier) yeah it's not really a must-have; it's just we have many variations of `InvalidRequest` so I thought it might worth to have an extra check, but can remove it if it looks excessive.

Reply:
My general rule here is that tests test functionality, what the error says doesn't matter as much as an error was thrown ( or in this case returned ). If the specific error was part of the functionality than turning it into some sort of object comparison would be ideal. A typo or updated message should not break a test basically.

Reply:
That makes sense and I agree in the general context, although for our case, the error types are very generic (I.e. InvalidRequest or InvalidArgument) that returning the correct type doesn’t mean much/enough; in a sense, the correct error description is part of the functionality, so I thought it might still worth to keep it to verify the behavior by checking that we’re actually sending the right messages.

Reply:
hmm I agree that those error types are too generic, and definitely not very helpful. Would it be possible ( maybe in the future ) to create more specific errors to return? Like MissingObjectLockConfigError, if not an instance of the generic error + description we can use to compare?

Reply:
oh yeah I think that’s a great idea. The thing is, seems like we (unnecessarily) mimicked AWS logic internally as well and are only differentiating the errors on the description part, but, would be a lot better to have custom errors instead. I also see a lot of errors being returned with `.customizeDescription('some custom description')` on the fly with the exact same descriptions; in the same process, they could also be refactored to use that error type and extract the description from there which could be another improvement I think.

Reply:
hmm actually the moment I wrote this I realized we already have `arsenalErrors.json` which could serve for this. I guess I could just add those errors with customized descriptive names and descriptions and use from there; e.g:
```
"InvalidRetentionDate": {
    "code": 400,
    "description": "The retain until date must be in the future!"
  }
```
It might get bit out of time (due to Arsenal involvement as well) for that PR but I think it's worth as a follow-up afterward.

Reply:
Yup exactly the sort of stuff I was thinking, will resolve this comment but we should do something like that.